Base/Python/slicer/parameterNodeWrapper/guiConnectors.py

```diff
@@ -0,0 +1,709 @@
+import abc
+import enum
+import pathlib
+import logging
+
+import ctk
+import qt
+
+import slicer
+from qSlicerSubjectHierarchyModuleWidgetsPythonQt import qMRMLSubjectHierarchyTreeView
+from . import parameterPack as pack
+from . import validators
+from .util import (
+    findFirstAnnotation,
+    getNodeTypes,
+    isNodeOrUnionOfNodes,
+    splitAnnotations,
+    unannotatedType,
+)
+
+
+__all__ = [
+    "parameterNodeGuiConnector",
+    "GuiConnector",
+    "SlicerPackParameterNamePropertyName",
+]
+
+
+class GuiConnector(abc.ABC):
+    """
+    Base class for converting from widgets to a datatype.
+    """
+
+    @staticmethod
+    @abc.abstractmethod
+    def canRepresent(widget, datatype) -> bool:
+        """
+        Whether this type can represent the given datatype using the given widget.
+        The datatype may be annotated.
+        """
+        raise NotImplementedError("Must implement canRepresent")
+
+    @staticmethod
+    @abc.abstractmethod
+    def create(widget, datatype):
+        """
+        Creates a new connector adapting the given widget object to the given (possibly annotated) datatype.
+        """
+        raise NotImplementedError("Must implement create")
+
+    def __init__(self):
+        super().__init__()
+        self._callback = None
+
+    def onChanged(self, callback):
+        self._callback = callback
+        if self._callback is None:
+            # disconnect to remove the reference on this connector from
+            # widget so this connector can possibly be garbage collected.
+            self._disconnect()
+        else:
+            self._connect()
+
+    def changed(self):
+        if self._callback:
+            self._callback()
+
+    @abc.abstractmethod
+    def _connect(self):
+        """
+        Make the necessary connection(s) to the widget.
+        """
+        raise NotImplementedError("Must implement _connect")
+
+    @abc.abstractmethod
+    def _disconnect(self):
+        """
+        Make the necessary disconnection(s) to the widget.
+        """
+        raise NotImplementedError("Must implement _disconnect")
+
+    @abc.abstractmethod
+    def widget(self):
+        """
+        Returns the underlying widget.
+        """
+        raise NotImplementedError("Must implement widget")
+
+    @abc.abstractmethod
+    def read(self):
+        """
+        Returns the value from the widget as the given datatype.
+        """
+        raise NotImplementedError("Must implement read")
+
+    @abc.abstractmethod
+    def write(self, value) -> None:
+        """
+        Writes the given value to the widget.
+        """
+        raise NotImplementedError("Must implement write")
+
+
+_registeredGuiConnectors = []
+
+
+def _processGuiConnector(classtype):
+    if not issubclass(classtype, GuiConnector):
+        raise TypeError("Must be a GuiConnector subclass")
+    global _registeredGuiConnectors
+    _registeredGuiConnectors.append(classtype)
+    return classtype
+
```

Comment:
In my mind, there are 3 major components here:

1. The abstract GuiConnector
2. The concrete GuiConnectors
3. The factory

Right now the dependencies are: factory -> concrete -> abstract

I feel like combining the abstract and the factory would make a cycle between these concepts: factory/abstract <-> concrete

Also, I have concerns about being able to do `ConcreteGuiConnector.registerGuiConnector/create` and it being allowed.

Reply:
I am somewhat against the implicit magic using `__subclasses__`. The decorator is explicit and descriptive.

Reply:
> I am somewhat against the implicit magic using `__subclasses__`. The decorator is explicit and descriptive.

This is fair. Relying on `__init_subclass__` has a similar issue. Given the point above about inheriting that registration method, it is probably better to keep the decorator even if it is doing "more work" than is strictly required.

---

> In my mind, there are 3 major components here:
> 
> - The abstract GuiConnector
> - The concrete GuiConnectors
> - The factory
>
> Right now the dependencies are: factory -> concrete -> abstract

Each concrete class already implements its own factory, `create`. My interpretation is that `cls.create()` should create a connector which is an instance of `cls`, different from a constructor.

I see the current dependency more like this, where the public factory function accesses the concrete factory implementation via that interface.

```
concrete --> abstract
   |           |
 create  <-- create ---- factory
```

`GuiConnector.create` then seems an appropriate place to create a guiconnector of any registered type.

I suppose the real point is: what's the difference in purpose between staticmethod `create` and `__init__()`? What's the difference between `createGuiConnector(...)` and a hypothetical `GuiConnector(...)`, or a hypothetical `GuiConnector.create(...)`?

To me, it is that `__init__`/`__new__` may only create an instance of the given type. `create` should _find_ an appropriate type, and create an instance of that.

However as things are, there does not seem much meaningful distinction between the concrete implementations of `create()` and their constructors. `createGuiConnector` already calls `canRepresent`, so does `create()` really need to? If that were removed, `create` truly would defer to the constructor.

Reply:
> what's the difference in purpose between staticmethod `create` and `__init__()`?

Roughly the difference is in levels of abstraction at which each should be used. `create` has a consistent interface across all derived classes. The constructor's interface is _what makes sense for the class_, while `create`'s interface is _what is general for the interface_.

I have reservations that I cannot fully articulate about requiring all the constructors to have the same interface. Maybe it is because I cannot define that interface in the base class.

The other thing separating `create` from `__init__` allows is the ability to special case special things that may not fit into the general interface (not necessarily recommended to do this, but I wanted a way if you really need to).

E.g.

```py
@parameterNodeGuiConnector
class SpecialGuiConnector:
  @staticmethod
  def create(widget, datatype):
    return SpecialGuiConnector(widget, datatype)

  def __init__(self, widget, datatype, someSpecialFlag=False):
    # impl

  # rest of interface

# explicit mapping that doesn't go through the general factory.
parameterNodeWrapper.connectParametersToGui({
  # the value here can be a specific connector to use rather than the widget
  "someParam": SpecialGuiConnector(someParamWidget, someType, True)
})
```

So while "there does not seem much meaningful distinction between the concrete implementations of create() and their constructors" right now for standard connectors, there could be for custom ones.

> I see the current dependency more like this, where the public factory function accesses the concrete factory implementation via that interface.

You are correct, that was a bad characterization on my part. The factory depends on the interface, but is semantically coupled to the implementation (it produces implementations as-a interface).

I am very strongly against "a hypothetical `GuiConnector(...)`" because, as you said, the constructor should only create its own type, but I don't think you are advocating for it.

In my view of the world, a factory function for a class inside that class should only produce instances of that particular class, possibly virtually (like `vtkObject::NewInstance`), but without logic that produces something outside your inheritance tree. A separate factory function outside the class it is producing (could be a factory class, but the important thing is the factory does not return Self type) may produce instances of anything derived from its associated "produced" class.

The concepts of "choose the best option style factory" and "how do I make my particular class" are so very different that I believe they shouldn't be put together in one class.

It is typical to use in-class factories to essentially create named constructors or virtual constructors, which are both semantically different than "choose the best derived class from a whole tree with many branches".

Here are a couple other "you shouldn't be able to do that" style things that having the create and register functions in class would allow:

```py
# this is very confusing as to what will happen if you don't know the underlying code
DerivedConnectorClass.registerGuiConnector(OtherDerivedConnector)
# the connector instance should not be a factory. What happens is dependent on the type of the instance.
connectorInstance.registerGuiConnector(OtherDerivedConnector)
# the connector instance should not be a factory. What happens is dependent on the type of the instance.
connectorInstance.create(randomWidget, randomDatatype)
```

Reply:
> connectorInstance.create(randomWidget, randomDatatype)

This behavior is already present.

However I do agree that those examples of `registerGuiConnector` are... dubious.

> Maybe it is because I cannot define that interface in the base class. [...] special things that may not fit into the general interface
> 
> A separate factory function outside the class it is producing (could be a factory class, but the important thing is the factory does not return Self type) may produce instances of anything derived from its associated "produced" class.

I suppose my main contention is that `create` feels more coupled to the factory mechanism, so they should either act like a factory, or simply use the constructor instead since the constructor (well, `__new__` and `__init__` together) _are_ a factory for a given class.

I wonder if it would be better to somehow let the factory mechanism reside _entirely_ outside the class, (`createGuiConnector`, `create`, and `canValidate`), however I'm not sure if there's a great way to do this without being overly verbose.

---

I'll think on it more but I would rather avoid bikeshedding an issue that does't affect the usability of things too much. If I or anyone else doesn't have an idea on how to better encapsulate these things I'd be okay with merging the current pattern.

Reply:
@Connor-Bowley one thought I've had is to move the `widget` and `datatype` tests to be declared in the `@parameterNodeGuiConnector` decorator.

Since most cases simply test if the arguments are one of a few types, simply borrow `isinstance` semantics.

```python
@parameterNodeGuiConnector(widget=(qt.QSlider, qt.QSpinBox), datatype=int)
class QSliderOrSpinBoxToIntConnector(GuiConnector):
    ...
```

To handle arbitrary tests, let `widget=` or `datatype=` arguments be a callable.

```python
@parameterNodeGuiConnector(
    widget=slicer.qMRMLNodeComboBox, datatype=isNodeOrUnionOfNodes
)
class qMRMLNodeComboBoxToNodeConnector(GuiConnector):
    ...
```

To handle non-standard constructor signatures, simply decorate a factory function instead.

```python
class NonstandardConnector(GuiConnector):
    def __init__(self, some, other, signature):
        ...

    ...


@parameterNodeGuiConnector(widget=qt.QSlider, datatype=float)
def createNonstandardConnector(widget, datatype):
    return NonstandardConnector(...)
```

This also has the benefit of supporting a single class for common interface among multiple connectors, by registering multiple factory functions for the same type.

```python
class ToggleToBoolConnector(GuiConnector):
    def __init__(self, widget, slot, datatype):
        self._widget = widget
        self._slot = slot
        self._datatype = datatype

    def _connect(self):
        self._slot.connect(self.changed)

    def _disconnect(self):
        self._slot.disconnect(self.changed)

    ...

@parameterNodeGuiConnector(widget=qt.QCheckBox, datatype=bool)
def QCheckBoxToBoolConnector(widget, datatype):
    return ToggleToBoolConnector(widget, widget.stateChanged, datatype)

@parameterNodeGuiConnector(widget=qt.QPushButton, datatype=bool)
def QPushButtonToBoolConnector(widget, datatype):
    if not widget.checkable:
        logging.warn('...')
        widget.checkable = True

    return ToggleToBoolConnector(widget, widget.toggled, datatype)
```

See https://gist.github.com/allemangD/aa95afd34c17b258c33f50c43ae2e33e for an (untested) prototype implementation of this signature for `@parameterNodeGuiConnector` and `createGuiConnector(widget, datatype)`